### ISSUE TITLE
V8 compilation error on PPC

### ISSUE THREAD

**Issue by targos** [nodejs/node-v8#119]
See: https://ci.nodejs.org/job/node-test-commit-v8-linux/2590/nodes=ppcle-ubuntu1404,v8test=v8test/console

```
08:07:55 cc1plus: error: unrecognized command line option '-Wno-class-memaccess' [-Werror]
08:07:55 cc1plus: error: unrecognized command line option '-Wno-packed-not-aligned' [-Werror]
08:07:55 cc1plus: all warnings being treated as errors
```

@nodejs/platform-ppc

**Comment by miladfarca** [comment]
It seems like `v8_enable_backtrace = true` is missing from `GN` flags, could you try to add this flag and recompile V8.

**Comment by miladfarca** [comment]
Flag should be added to the Jenkins job which runs `gn`. The above console output shows the current flag is set to the following:
```
gn gen out.gn/ppc64.release '--args=is_component_build=false is_debug=false use_goma=false goma_dir="None" use_custom_libcxx=false v8_target_cpu="ppc64" target_cpu="ppc64"'
```
`v8_enable_backtrace = true` can be be added to the args list.

**Comment by targos** [comment]
Trying: https://ci.nodejs.org/job/node-test-commit-v8-linux/2594/

**Comment by targos** [comment]
It builds, but hits another error: 

```
08:57:39 [1479/1479] LINK ./cctest
08:57:39 deps/v8/tools/run-tests.py --gn --arch=ppc64 \
08:57:39 				--mode=release --progress=dots --timeout=120 \
08:57:39 				mjsunit cctest debugger inspector message preparser \
08:57:39 				--junitout /home/iojs/build/workspace/node-test-commit-v8-linux/nodes/ppcle-ubuntu1404/v8test/v8test/v8-tap.xml
08:57:40 Traceback (most recent call last):
08:57:40   File "/home/iojs/build/workspace/node-test-commit-v8-linux/nodes/ppcle-ubuntu1404/v8test/v8test/deps/v8/tools/testrunner/base_runner.py", line 281, in execute
08:57:40     tests = self._load_testsuite_generators(args, options)
08:57:40   File "/home/iojs/build/workspace/node-test-commit-v8-linux/nodes/ppcle-ubuntu1404/v8test/v8test/deps/v8/tools/testrunner/base_runner.py", line 651, in _load_testsuite_generators
08:57:40     self.framework_name)
08:57:40   File "/home/iojs/build/workspace/node-test-commit-v8-linux/nodes/ppcle-ubuntu1404/v8test/v8test/deps/v8/tools/testrunner/local/testsuite.py", line 254, in Load
08:57:40     with _load_testsuite_module(name, root) as module:
08:57:40   File "/usr/lib/python2.7/contextlib.py", line 17, in __enter__
08:57:40     return self.gen.next()
08:57:40   File "/home/iojs/build/workspace/node-test-commit-v8-linux/nodes/ppcle-ubuntu1404/v8test/v8test/deps/v8/tools/testrunner/local/testsuite.py", line 245, in _load_testsuite_module
08:57:40     yield imp.load_module(name + "_testcfg", f, pathname, description)
08:57:40 SyntaxError: unqualified exec is not allowed in function '_ParsePythonTestTemplates' it contains a nested function with free variables (testcfg.py, line 71)
```

**Comment by miladfarca** [comment]
I checked out `node-v8`, fetched dependencies according the the `console` output and tests start normally.

I did however get the above error on a V8 checkout with older dependencies, running "gclient sync" fixed it.

Could you try rerunning  a fresh Jenkins job? I also don't have access to thar machine, could you  please add my key for any further investigations, I got permission here: https://github.com/nodejs/build/issues/1706
```
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDQZihjEXcY52UZo09CEb29HwOWwXcmwbwZFR4rsACQQyGUexL4fkVrFfwuG8eV1vg5KECsO8LiGY/MHkNIpABaJJoip0Qxgv0pAFtAukIDjLXXOV/VNJjfIto16vOAehRZkmI+BtQP8TjoT2CSyJgvVQcay8BhH52in1LQQsyCi2crHLYzDrrCgY/rAmuVb1MzMnT8mFOdJ8E5RBhjnmc1K4YBKmNTf6yefgbOJssI0lLzp7Q2uytzp3pipg7AO/VqmRn8953UTJS/cOQeBi3nCYGpz4I7kOKHgwbdW1IP/XFfm0KO5daulHQeToRGIE85ntxF314wsYE3ZyeJwKgH
```

**Comment by sam-github** [comment]
Anyone with @nodejs/build access can add authorized keys to test machines (if an issue exists to say its OK!). I added @miladfarca.

**Comment by miladfarca** [comment]
There seems to be a bug in Python versions prior to `2.7.9` which produces the above error:
https://bugs.python.org/issue21591
https://github.com/pyinstaller/pyinstaller/issues/1408
https://github.com/django-extensions/django-extensions/issues/1219

Python version on this machine is `2.7.6`. This is the V8 CL which has introduced this issue:
https://chromium-review.googlesource.com/c/v8/v8/+/1864942

I have asked the owner for a possible workaround (or python has to be upgraded/patched).

**Comment by sam-github** [comment]
I can schedule python updates if it ends up that is what's necessary. It'll require updates to all the ansilbe scripts that build python for those hosts (likely, unless some are getting it from package repos with updates, but thats not likely).

I'd need an explicit list of hosts that need the update - all `*ppc64*`? just ubuntu? just centos? Its not clear ATM.

Also, I'd need a minimum python2 version.

**Comment by miladfarca** [comment]
Thanks Sam, 
I have proposed a change for a workaround in this CL, waiting for a review:
https://chromium-review.googlesource.com/c/v8/v8/+/1893647

Let's see if they accept the change considering Python 3 is going to replace it sometime very soon. If not then we might have to upgrade Python to at-least `2.7.9` (preferably `2.7.15+`).

**Comment by miladfarca** [comment]
above CL is now landed. Issue should be solved on the next lkgr build.

**Comment by targos** [comment]
Thanks.

Before closing this, can I ask why `v8_enable_backtrace=true` fixed the issue?
Should we also set it in our .gyp files to build Node.js?

**Comment by miladfarca** [comment]
We always had this flag in our V8 internal builds, I feel this might be a gcc limitation where not having the backtrace symbols causes lambda related syntax errors. More info can be found here:
https://www.gnu.org/software/libc/manual/html_node/Backtraces.html

**Comment by miladfarca** [comment]
This is CL is now merged https://github.com/v8/v8/commit/0c3906f4dce0aadf3675687837642aa838c80567 however it's breaking Python 3 compatibility as @cclauss has mentioned and I will have to revert it.

@sam-github The only way to get around this bug is to upgrade our python versions from `2.7.6` to
any version above `2.7.9`. I have tested `2.7.15+` and it works fine, was wondering if you could help schedule python updates.
@mhdawson Would you be able to provide a list of machines you want updated, which have python `< 2.7.9` as default. 

I have seen the above bug on both `test-osuosl-centos7-ppc64_le-1` and `test-osuosl-ubuntu1404-ppc64_le-1`. I also checked the python version on `test-linuxonecc-rhel72-s390x-1` and it's `2.7.5` (needs to be updated).

Keep in mind this is only needed on machines that compile and run `V8` tests.

**Comment by cclauss** [comment]
As discussed in nodejs/build#1802

**Comment by sam-github** [comment]
Is reworking https://github.com/v8/v8/commit/0c3906f4dce0aadf3675687837642aa838c80567 so that it works for py2 and py3 not a much easier option, and one which would increase the robustness of the v8 built scripts?

I checked the three machines you listed, they all use latest package python2 from the distro, so in order to update we'll have to abandon using pre-packaged binaries and build python from source.

> Keep in mind this upgrade is only needed on machines that compile and run V8 tests.

So just the three you listed?

**Comment by miladfarca** [comment]
The above Cl is the only workaround possible from a syntax point of view which is not compatible with py3. This is valid py2 syntax, it's just that our python versions are too old and need to be patched.
There might be more of such commits as they are preparing v8 to use with py3, seems like we might need to patch some of our machines.

These are the only 3 machines I noticed on the job, there could be more which Michael might be aware of.

**Comment by sam-github** [comment]
@miladfarca I think you only found one single syntax that works for both py3 and py2.7.9+, but I'm suggesting a conditional: if py3 do X, if py2 do Y.

As you point out, google is going towards python3, so time spent trying to update python2 seems a bit  pointless.

We'll do the update if its blocking, but no packages exist, so its a matter of building from source, and its necessary to find all the -devel packages so that the required python features are present, and we don't have a list of those features, so worst case might require iteration. See https://github.com/nodejs/node/issues/30152#issuecomment-552554464, which was easier because the platform was already building python, it just needed a tweek (two tweeks, actually).

**Comment by sam-github** [comment]
I guess this issue needs to be reopened?

**Comment by mhdawson** [comment]
> @mhdawson Would you be able to provide a list of machines you want updated, which have python < 2.7.9 as default.

@sam-github should be able to generate that list.

If ' if py3 do X, if py2 do Y.' is an option then that will save us a bunch of time (unless of course we later have to upgrade for some other reason).

**Comment by miladfarca** [comment]
The problem is python will byte-compile your source file before starting to execute it. The whole file must at least parse correctly. My previous patch isn't a correct syntax in python 3 so something like this:
```python
if sys.version_info >= (3, 0):
    exec("print(dir())", {}) 
else:
    exec "print(dir())" in {}
```
won't even run with py 3 as `exec "print(dir())" in {}` is invalid syntax (but the only workaround possible to the above python2 bug).

**Comment by cclauss** [comment]
I have seen other projects deal with this using
```python
if sys.version_info >= (2, 7, 9):
    import exec_command_modern as exec_command
else:
    import exec_command_legacy as exec_command

exec_command()
```

However, my strong preference would be that we drop support for Python < 2.7.9.

**Comment by miladfarca** [comment]
I have created a new CL which uses a combination of `eval` and `compile`. Doesn't look pretty but it
works in both py2 and 3. Waiting for it to be reviewed:
https://chromium-review.googlesource.com/c/v8/v8/+/1925028

**Comment by sam-github** [comment]
Should we be opening another issue to get the V8 builds to use python3?

I'm not very familiar with the V8 build jobs:

Do they support python 3? I've seen some comments which I can't find ATM that indicate that the google tools might not all be py3 capable yet.

Do they auto-detect python, and prefer python3 if found? If so, we could just install python 3 on some set of machines -- "install" via adding ansible configuration to build python 3 from source)

If they don't auto-detect, what has to be done? Should we just be waiting for auto-detect and prefer of py3 to be complete?

**Comment by targos** [comment]
V8 builds fail with Python 3, because Google tooling is not compatible.
See https://ci.nodejs.org/job/node-test-commit-v8-linux/2640/ for example.

**Comment by sam-github** [comment]
OK, in that case, patching as little as possible until they support python3, then moving to python3 would be my preferred path. If we can't get them to be a bit more lenient with the patch version of 2.7 they support, then we'll be forced to move away from distro-packaged Python.

**Comment by miladfarca** [comment]
V8 community has now removed the entire "preparser" (which includes the line we are changing) in favour of "cctest/test-parsing": 
https://chromium-review.googlesource.com/c/v8/v8/+/1926147
